File: spec/system/sla_tracking_spec.rb
# Migrated from test/system/sla_tracking_test.rb
# System spec: SLA tracking and workflow deadline monitoring

require 'rails_helper'

RSpec.describe "SLA Tracking System", type: :system do
  # Critical Business Path: SLA monitoring and deadline tracking
  # Risk Level: HIGH - SLA violations affect customer service and compliance
  # Focus: Visual indicators, deadline calculations, and workflow status tracking

  before do
    setup_authenticated_session(role: "manager", company_id: 1)
    setup_sla_test_data

    # Mock time for consistent SLA calculations
    allow(Time).to receive(:current).and_return(Time.parse("2024-01-15 10:00:00 UTC"))
  end

  describe "SLA indicators on invoices index page" do
    it "displays correct SLA status indicators for different invoice states" do
      # Mock invoice list with SLA data
      allow(InvoiceService).to receive(:all).and_return({
        invoices: [
          @invoice_with_sla,
          @overdue_invoice,
          @invoice_without_sla
        ],
        meta: { total: 3, page: 1, pages: 1 }
      })

      visit invoices_path

      # Check normal SLA indicator (green - 4 hours remaining)
      within "tr", text: "INV-001" do
        expect(page).to have_text("Due in 4 hours")
        expect(page).to have_css(".text-green-600.bg-green-100")
      end

      # Check overdue SLA indicator (red - 1 hour overdue)
      within "tr", text: "INV-002" do
        expect(page).to have_text("Overdue by 1 hour")
        expect(page).to have_css(".text-red-600.bg-red-100")
      end

      # Check invoice without SLA (gray)
      within "tr", text: "INV-004" do
        expect(page).to have_text("No SLA")
        expect(page).to have_css(".text-gray-500.bg-gray-100")
      end
    end

    it "displays warning indicator for approaching deadlines" do
      allow(InvoiceService).to receive(:all).and_return({
        invoices: [@warning_invoice],
        meta: { total: 1, page: 1, pages: 1 }
      })

      visit invoices_path

      within "tr", text: "INV-003" do
        expect(page).to have_text("Due in 30 minutes")
        expect(page).to have_css(".text-yellow-600.bg-yellow-100")
      end
    end

    it "applies correct styling based on time remaining" do
      allow(InvoiceService).to receive(:all).and_return({
        invoices: [
          @invoice_with_sla,
          @overdue_invoice,
          @warning_invoice
        ],
        meta: { total: 3, page: 1, pages: 1 }
      })

      visit invoices_path

      # Normal SLA (green)
      within "tr", text: "INV-001" do
        expect(page).to have_css(".text-green-600.bg-green-100")
      end

      # Overdue SLA (red)
      within "tr", text: "INV-002" do
        expect(page).to have_css(".text-red-600.bg-red-100")
      end

      # Warning SLA (yellow)
      within "tr", text: "INV-003" do
        expect(page).to have_css(".text-yellow-600.bg-yellow-100")
      end
    end
  end

  describe "detailed SLA information on workflow page" do
    it "displays comprehensive SLA details for active invoices" do
      allow(InvoiceService).to receive(:find).and_return(@invoice_with_sla)
      allow(WorkflowService).to receive(:available_transitions).and_return({
        available_transitions: []
      })
      allow(WorkflowService).to receive(:history).and_return([])

      visit invoice_workflow_path(@invoice_with_sla[:id])

      # Check SLA status section
      expect(page).to have_text("SLA Status:")
      expect(page).to have_text("Due in 4 hours")

      # Check detailed SLA section
      expect(page).to have_text("SLA Details")
      expect(page).to have_text("Time in current state: 2 hours")
      expect(page).to have_text("Deadline: Jan 15, 2024 at 02:00 PM")

      # Check progress bar presence
      expect(page).to have_css(".bg-green-500")
    end

    it "displays overdue SLA details with correct styling" do
      allow(InvoiceService).to receive(:find).and_return(@overdue_invoice)
      allow(WorkflowService).to receive(:available_transitions).and_return({
        available_transitions: []
      })
      allow(WorkflowService).to receive(:history).and_return([])

      visit invoice_workflow_path(@overdue_invoice[:id])

      # Check overdue status
      expect(page).to have_text("Overdue by 1 hour")
      expect(page).to have_css(".text-red-600.bg-red-100")

      # Check detailed SLA section shows progress as overdue
      expect(page).to have_text("SLA Details")
      expect(page).to have_css(".bg-red-500") # Progress bar should be red for overdue
    end

    it "handles invoices without SLA deadline" do
      allow(InvoiceService).to receive(:find).and_return(@invoice_without_sla)
      allow(WorkflowService).to receive(:available_transitions).and_return({
        available_transitions: []
      })
      allow(WorkflowService).to receive(:history).and_return([])

      visit invoice_workflow_path(@invoice_without_sla[:id])

      # Should show "No SLA" status
      expect(page).to have_text("No SLA")
      expect(page).to have_css(".text-gray-500.bg-gray-100")

      # Should not show detailed SLA section
      expect(page).not_to have_text("SLA Details")
    end

    it "calculates and displays correct progress percentage" do
      # Invoice that's 50% through its SLA period
      halfway_invoice = {
        id: 5,
        invoice_number: "INV-005",
        status: "pending_review",
        workflow: {
          'entered_current_state_at' => '2024-01-15 08:00:00 UTC', # 2 hours ago
          'sla_deadline' => '2024-01-15 12:00:00 UTC', # 2 hours from now (4 hours total)
          'is_overdue' => false
        }
      }

      allow(InvoiceService).to receive(:find).and_return(halfway_invoice)
      allow(WorkflowService).to receive(:available_transitions).and_return({
        available_transitions: []
      })
      allow(WorkflowService).to receive(:history).and_return([])

      visit invoice_workflow_path(halfway_invoice[:id])

      # Progress bar should be present for 50% completion
      progress_bar = find(".bg-green-500")
      expect(progress_bar).to be_present
    end
  end

  describe "edge cases and error handling" do
    it "handles missing workflow data gracefully" do
      invoice_no_workflow = {
        id: 6,
        invoice_number: "INV-006",
        status: "draft"
        # No workflow key
      }

      allow(InvoiceService).to receive(:find).and_return(invoice_no_workflow)
      allow(WorkflowService).to receive(:available_transitions).and_return({
        available_transitions: []
      })
      allow(WorkflowService).to receive(:history).and_return([])

      visit invoice_workflow_path(invoice_no_workflow[:id])

      # Should not show SLA section at all
      expect(page).not_to have_text("SLA Status:")
      expect(page).not_to have_text("SLA Details")
    end
  end

  private

  def setup_sla_test_data
    @invoice_with_sla = {
      id: 1,
      invoice_number: "INV-001",
      status: "pending_review",
      workflow: {
        'entered_current_state_at' => '2024-01-15 08:00:00 UTC',
        'sla_deadline' => '2024-01-15 14:00:00 UTC',
        'is_overdue' => false
      }
    }

    @overdue_invoice = {
      id: 2,
      invoice_number: "INV-002",
      status: "pending_review",
      workflow: {
        'entered_current_state_at' => '2024-01-14 08:00:00 UTC',
        'sla_deadline' => '2024-01-15 09:00:00 UTC',
        'is_overdue' => true
      }
    }

    @warning_invoice = {
      id: 3,
      invoice_number: "INV-003",
      status: "pending_review",
      workflow: {
        'entered_current_state_at' => '2024-01-15 08:00:00 UTC',
        'sla_deadline' => '2024-01-15 10:30:00 UTC', # 30 minutes from now
        'is_overdue' => false
      }
    }

    @invoice_without_sla = {
      id: 4,
      invoice_number: "INV-004",
      status: "draft",
      workflow: {
        'entered_current_state_at' => '2024-01-15 08:00:00 UTC'
        # No sla_deadline
      }
    }

    # Mock common services
    allow(InvoiceService).to receive(:recent).and_return([])
  end
end

File: e2e/tests/workflows/sla-tracking.spec.ts
import { test, expect } from '../../fixtures/auth.fixture';
import { WorkflowPage } from '../../pages/workflow.page';
import { Page } from '@playwright/test';

// Mock time for consistent SLA calculations
const MOCK_CURRENT_TIME = new Date('2024-01-15T10:00:00Z');

test.describe('SLA Tracking System', () => {
  let workflowPage: WorkflowPage;

  test.beforeEach(async ({ authenticatedPage }) => {
    workflowPage = new WorkflowPage(authenticatedPage);

    // Mock time for consistent SLA calculations
    await authenticatedPage.addInitScript(() => {
      const mockDate = new Date('2024-01-15T10:00:00Z');
      Date.now = () => mockDate.getTime();
      Date.prototype.getTime = function() { return mockDate.getTime(); };
    });

    // Setup API mocks with SLA data
    await setupSLAMocks(authenticatedPage);
  });

  test.describe('SLA indicators on invoices index page', () => {
    test('displays correct SLA status indicators for different invoice states', async () => {
      await workflowPage.gotoInvoicesList();

      // Check normal SLA indicator (green - 4 hours remaining)
      const normalSLA = await workflowPage.getInvoiceSLAIndicator('INV-001');
      expect(normalSLA.text).toContain('Due in 4 hours');
      expect(normalSLA.color).toBe('green');

      // Check overdue SLA indicator (red - 1 hour overdue)
      const overdueSLA = await workflowPage.getInvoiceSLAIndicator('INV-002');
      expect(overdueSLA.text).toContain('Overdue by 1 hour');
      expect(overdueSLA.color).toBe('red');

      // Check invoice without SLA (gray)
      const noSLA = await workflowPage.getInvoiceSLAIndicator('INV-004');
      expect(noSLA.text).toContain('No SLA');
      expect(noSLA.color).toBe('gray');
    });

    test('displays warning indicator for approaching deadlines', async () => {
      await workflowPage.gotoInvoicesList();

      // Check warning SLA indicator (yellow - 30 minutes remaining)
      const warningSLA = await workflowPage.getInvoiceSLAIndicator('INV-003');
      expect(warningSLA.text).toContain('Due in 30 minutes');
      expect(warningSLA.color).toBe('yellow');
    });

    test('applies correct styling based on time remaining', async () => {
      await workflowPage.gotoInvoicesList();

      const allIndicators = await workflowPage.getAllInvoiceSLAIndicators();

      // Find specific invoices and verify their styling
      const inv001 = allIndicators.find(i => i.invoiceNumber.includes('INV-001'));
      const inv002 = allIndicators.find(i => i.invoiceNumber.includes('INV-002'));
      const inv003 = allIndicators.find(i => i.invoiceNumber.includes('INV-003'));

      expect(inv001?.slaColor).toBe('green'); // Normal
      expect(inv002?.slaColor).toBe('red');   // Overdue
      expect(inv003?.slaColor).toBe('yellow'); // Warning
    });
  });

  test.describe('detailed SLA information on workflow page', () => {
    test('displays comprehensive SLA details for active invoices', async () => {
      await workflowPage.gotoInvoiceWorkflow('1');

      // Check SLA status section
      const status = await workflowPage.getSLAStatus();
      expect(status.message).toContain('Due in 4 hours');
      expect(status.color).toBe('green');

      // Check detailed SLA section
      const details = await workflowPage.getSLADetails();
      expect(details).not.toBeNull();
      expect(details!.timeInState).toContain('2 hours');
      expect(details!.deadline).toContain('Jan 15, 2024');
      expect(details!.deadline).toContain('14:00'); // Should be 2:00 PM
      expect(details!.isOverdue).toBe(false);
      expect(details!.progress).toBeGreaterThanOrEqual(0);
      expect(details!.progress).toBeLessThanOrEqual(100);
    });

    test('displays overdue SLA details with correct styling', async () => {
      await workflowPage.gotoInvoiceWorkflow('2');

      // Check overdue status
      const status = await workflowPage.getSLAStatus();
      expect(status.message).toContain('Overdue by 1 hour');
      expect(status.color).toBe('red');

      // Check detailed SLA section shows as overdue
      const details = await workflowPage.getSLADetails();
      expect(details).not.toBeNull();
      expect(details!.isOverdue).toBe(true);
    });

    test('handles invoices without SLA deadline', async () => {
      await workflowPage.gotoInvoiceWorkflow('4');

      // Should show "No SLA" status
      const status = await workflowPage.getSLAStatus();
      expect(status.message).toContain('No SLA');
      expect(status.color).toBe('gray');

      // Should not show detailed SLA section
      const details = await workflowPage.getSLADetails();
      expect(details).toBeNull();
    });

    test('calculates and displays correct progress percentage', async () => {
      // Mock invoice that's 50% through its SLA period
      await workflowPage.gotoInvoiceWorkflow('5');

      const details = await workflowPage.getSLADetails();
      expect(details).not.toBeNull();

      // Progress should be around 50% (2 hours passed out of 4 hours total)
      expect(details!.progress).toBeGreaterThanOrEqual(45);
      expect(details!.progress).toBeLessThanOrEqual(55);
    });
  });

  test.describe('edge cases and error handling', () => {
    test('handles missing workflow data gracefully', async () => {
      // Mock invoice without workflow data
      await workflowPage.page.route('**/api/v1/invoices/6', async route => {
        await route.fulfill({
          status: 200,
          contentType: 'application/json',
          body: JSON.stringify({
            id: '6',
            invoice_number: 'INV-006',
            status: 'draft',
            // No workflow key
          }),
        });
      });

      await workflowPage.gotoInvoiceWorkflow('6');

      // Should not show SLA section at all
      await expect(workflowPage.slaStatusBadge).not.toBeVisible();
      await expect(workflowPage.slaDetailsSection).not.toBeVisible();
    });

    test('updates SLA status after workflow transition', async () => {
      await workflowPage.gotoInvoiceWorkflow('1');

      // Get initial status
      const initialStatus = await workflowPage.getSLAStatus();
      expect(initialStatus.color).toBe('green');

      // Mock transition response with new SLA
      await workflowPage.page.route('**/api/v1/workflows/transition', async route => {
        await route.fulfill({
          status: 200,
          contentType: 'application/json',
          body: JSON.stringify({
            success: true,
            workflow: {
              entered_current_state_at: new Date().toISOString(),
              sla_deadline: new Date(Date.now() + 1 * 60 * 60 * 1000).toISOString(), // 1 hour from now
              is_overdue: false,
            },
          }),
        });
      });

      // Perform transition if available
      const transitions = await workflowPage.getAvailableTransitions();
      if (transitions.length > 0) {
        await workflowPage.performTransition(transitions[0]);
        await workflowPage.waitForSLAUpdate();

        // Check updated status
        const updatedStatus = await workflowPage.getSLAStatus();
        expect(updatedStatus.message).toContain('Due in');
      }
    });
  });
});

/**
 * Setup mock data for SLA testing
 */
async function setupSLAMocks(page: Page) {
  const apiUrl = process.env.API_URL || 'http://localhost:3000/api/v1';

  // Mock invoice list with SLA data
  await page.route(`${apiUrl}/invoices`, async route => {
    await route.fulfill({
      status: 200,
      contentType: 'application/json',
      body: JSON.stringify({
        invoices: [
          {
            id: '1',
            invoice_number: 'INV-001',
            status: 'pending_review',
            workflow: {
              entered_current_state_at: '2024-01-15T08:00:00Z', // 2 hours ago
              sla_deadline: '2024-01-15T14:00:00Z', // 4 hours from now
              is_overdue: false,
            },
          },
          {
            id: '2',
            invoice_number: 'INV-002',
            status: 'pending_review',
            workflow: {
              entered_current_state_at: '2024-01-14T08:00:00Z', // Yesterday
              sla_deadline: '2024-01-15T09:00:00Z', // 1 hour ago (overdue)
              is_overdue: true,
            },
          },
          {
            id: '3',
            invoice_number: 'INV-003',
            status: 'pending_review',
            workflow: {
              entered_current_state_at: '2024-01-15T08:00:00Z',
              sla_deadline: '2024-01-15T10:30:00Z', // 30 minutes from now
              is_overdue: false,
            },
          },
          {
            id: '4',
            invoice_number: 'INV-004',
            status: 'draft',
            workflow: {
              entered_current_state_at: '2024-01-15T08:00:00Z',
              // No sla_deadline
            },
          },
        ],
        meta: { total: 4, page: 1, pages: 1 },
      }),
    });
  });

  // Mock individual invoice endpoints
  const invoiceData = [
    {
      id: '1',
      invoice_number: 'INV-001',
      status: 'pending_review',
      workflow: {
        entered_current_state_at: '2024-01-15T08:00:00Z',
        sla_deadline: '2024-01-15T14:00:00Z',
        is_overdue: false,
      },
    },
    {
      id: '2',
      invoice_number: 'INV-002',
      status: 'pending_review',
      workflow: {
        entered_current_state_at: '2024-01-14T08:00:00Z',
        sla_deadline: '2024-01-15T09:00:00Z',
        is_overdue: true,
      },
    },
    {
      id: '3',
      invoice_number: 'INV-003',
      status: 'pending_review',
      workflow: {
        entered_current_state_at: '2024-01-15T08:00:00Z',
        sla_deadline: '2024-01-15T10:30:00Z',
        is_overdue: false,
      },
    },
    {
      id: '4',
      invoice_number: 'INV-004',
      status: 'draft',
      workflow: {
        entered_current_state_at: '2024-01-15T08:00:00Z',
      },
    },
    {
      id: '5',
      invoice_number: 'INV-005',
      status: 'pending_review',
      workflow: {
        entered_current_state_at: '2024-01-15T08:00:00Z', // 2 hours ago
        sla_deadline: '2024-01-15T12:00:00Z', // 2 hours from now (50% through 4-hour SLA)
        is_overdue: false,
      },
    },
  ];

  for (const invoice of invoiceData) {
    await page.route(`${apiUrl}/invoices/${invoice.id}`, async route => {
      await route.fulfill({
        status: 200,
        contentType: 'application/json',
        body: JSON.stringify(invoice),
      });
    });
  }

  // Mock workflow endpoints
  await page.route(`${apiUrl}/workflows/available_transitions`, async route => {
    await route.fulfill({
      status: 200,
      contentType: 'application/json',
      body: JSON.stringify({
        available_transitions: [
          { name: 'Approve', to_state: 'approved' },
          { name: 'Reject', to_state: 'rejected' },
        ],
      }),
    });
  });

  await page.route(`${apiUrl}/workflows/history`, async route => {
    await route.fulfill({
      status: 200,
      contentType: 'application/json',
      body: JSON.stringify([]),
    });
  });
}